File: rgn_99problems/src/main/java/ragna/kt/lists/P01.java
package ragna.kt.lists;

import ragna.ALists;

import java.util.List;
import java.util.NoSuchElementException;

public class P01 {

    static <T> T lastRecursive(List<T> list) {
        var head = ALists.head(list);
        var tail = ALists.tail(list);
        if (head != null && tail.isEmpty()) {
            return head;
        }
        if (!tail.isEmpty()) {
            return lastRecursive(tail);
        }
        throw new NoSuchElementException();
    }

    static <T> T last(List<T> list) {
        return lastRecursive(list);
    }

    public static void main(String[] args) {
        var list = List.of(1, 1, 2, 3, 5, 8);

        System.out.println(list);
        System.out.println(lastRecursive(list));
    }
}

File: rgn_99problems/src/main/kotlin/ragna/kt/lists/P01.kt
package ragna.kt.lists

import kotlinx.collections.immutable.ImmutableList
import kotlinx.collections.immutable.persistentListOf

fun <T> lastRecursive(list: List<T>): T {
    val head = list.head()
    val tail = list.tail()

    return when {
        head != null && tail.isEmpty() -> head
        tail.isNotEmpty() -> lastRecursive(tail)
        else -> throw NoSuchElementException()
    }
}

fun main() {
    val list: ImmutableList<Int> = persistentListOf(1, 1, 2, 3, 5, 8);
    println(list)
    println(lastRecursive(list))

}


fun <T> List<T>.tail(): List<T> {
    return drop(1)
}

fun <T> List<T>.head(): T {
    return first()
}
